src/ArrayHelper.php

```diff
@@ -1043,24 +1047,56 @@ public static function pathExists(
      */
     public static function htmlEncode(iterable $data, bool $valuesOnly = true, ?string $encoding = null): array
     {
-        $d = [];
+        if (!is_array($data)) {
+            $data = iterator_to_array($data);
+        }
+
+        $encoding ??= ini_get('default_charset') ?: 'UTF-8';
+        $flags = ENT_QUOTES | ENT_SUBSTITUTE;
+
+        return $valuesOnly
+            ? self::htmlEncodeValues($data, $encoding, $flags)
+            : self::htmlEncodeKeysAndValues($data, $encoding, $flags);
+    }
+
+    /**
+     * @param array $data
+     * @param string $encoding
+     * @param int $flags
+     * @return array
+     */
+    private static function htmlEncodeValues(array $data, string $encoding, int $flags): array
+    {
+        $result = [];
         foreach ($data as $key => $value) {
-            if (!is_int($key)) {
-                $key = (string)$key;
+            if (is_string($value)) {
+                $result[$key] = htmlspecialchars($value, $flags, $encoding, true);
+            } elseif (is_array($value)) {
+                $result[$key] = self::htmlEncodeValues($value, $encoding, $flags);
+            } else {
+                $result[$key] = $value;
             }
-            if (!$valuesOnly && is_string($key)) {
-                $key = htmlspecialchars($key, ENT_QUOTES | ENT_SUBSTITUTE, $encoding, true);
+        }
+        return $result;
+    }
+
+    private static function htmlEncodeKeysAndValues(array $data, string $encoding, int $flags): array
```

Comment:
Change order of parameters to `$data, $flags, $encoding` for consistency with `htmlspecialchars()`.